ray-curator/ray_curator/backends/experimental/ray_actor_pool/shuffle_adapter.py

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.
+"""Shuffle adapter for deduplication executor."""
+
+import math
+from typing import TYPE_CHECKING
+
+import ray
+from loguru import logger
+
+from ray_curator.backends.base import BaseStageAdapter
+from ray_curator.backends.experimental.utils import get_worker_metadata_and_node_id
+from ray_curator.tasks import FileGroupTask
+
+if TYPE_CHECKING:
+    from ray_curator.stages.deduplication.fuzzy.lsh.stage import LSHStage
+
+
+@ray.remote
+class ShuffleStageAdapter(BaseStageAdapter):
+    """Ray actor that wraps a shuffle stage and its actor.
+
+    This adapter manages the lifecycle of a shuffle actor (like LSHActor)
+    and provides a uniform interface for the executor.
+    """
+
+    def __init__(
+        self,
+        stage: "LSHStage",
+        rank: int,
+        nranks: int,
+        num_input_tasks: int | None = None,
+    ):
+        """Initialize the adapter.
+
+        Args:
+            stage: The shuffle stage to wrap
+            rank: This actor's rank in the group
+            nranks: Total number of actors in the group
+            session_id: Unique session identifier
+            input_nparts: Total input partitions
+        """
+        super().__init__(stage)
+        # Get runtime context for worker metadata (copied from RayActorPoolStageAdapter)
+        node_info, worker_metadata = get_worker_metadata_and_node_id()
+
+        # Create WorkerMetadata with actor information
+        self.worker_metadata = worker_metadata
+        self.node_info = node_info
+
+        self._batch_size = self.stage.batch_size
+        if self._batch_size is None:
+            logger.warning(f"batch size not set for stage {self.stage}. Setting it to 1.")
+            self._batch_size = 1
+
+        # Auto set total_nparts if not set
+        stage_class_kwargs = self.stage.actor_kwargs.copy()
+        if stage_class_kwargs.get("total_nparts") is None:
+            if num_input_tasks is None:
+                err_msg = "Shuffle total_nparts could not be set automatically. Please set it manually during stage initialization."
+                raise ValueError(err_msg)
+
+            self.output_nparts = max(1, 2 ** math.floor(math.log2(num_input_tasks)))
```

Comment:
Could you explain more where this calculation is coming from?

Reply:
This has been carried over from the old logic here: https://github.com/NVIDIA-NeMo/Curator/blob/ray-api/nemo_curator/modules/fuzzy_dedup/lsh.py#L182. 

The main motivation for this was to optimize the shuffle perf which is sensitive to num output partitions. The actual math is based more on empirical evidence where we could get away by rounding down to the nearest power of 2 (generally) without running into OOM issues at scale.